src/sema/assembly/builtin.rs

```diff
@@ -0,0 +1,654 @@
+    "sload" => AssemblyBuiltInFunction::SLoad,
+    "sstore" => AssemblyBuiltInFunction::SStore,
+    "msize" => AssemblyBuiltInFunction::MSize,
+    "gas" => AssemblyBuiltInFunction::Gas,
+    "address" => AssemblyBuiltInFunction::Address,
+    "balance" => AssemblyBuiltInFunction::Balance,
+    "selfbalance" => AssemblyBuiltInFunction::SelfBalance,
+    "caller" => AssemblyBuiltInFunction::Caller,
+    "callvalue" => AssemblyBuiltInFunction::CallValue,
+    "calldataload" => AssemblyBuiltInFunction::CallDataLoad,
+    "calldatasize" => AssemblyBuiltInFunction::CallDataSize,
+    "calldatacopy" => AssemblyBuiltInFunction::CallDataCopy,
+    "codesize" => AssemblyBuiltInFunction::CodeSize,
+    "codecopy" => AssemblyBuiltInFunction::CodeCopy,
+    "extcodesize" => AssemblyBuiltInFunction::ExtCodeSize,
+    "extcodecopy" => AssemblyBuiltInFunction::ExtCodeCopy,
+    "returndatasize" => AssemblyBuiltInFunction::ReturnDataSize,
+    "returndatacopy" => AssemblyBuiltInFunction::ReturnDataCopy,
+    "extcodehash" => AssemblyBuiltInFunction::ExtCodeHash,
+    "create" => AssemblyBuiltInFunction::Create,
+    "create2" => AssemblyBuiltInFunction::Create2,
+    "call" => AssemblyBuiltInFunction::Call,
+    "callcode" => AssemblyBuiltInFunction::CallCode,
+    "delegatecall" => AssemblyBuiltInFunction::DelegateCall,
+    "staticcall" => AssemblyBuiltInFunction::StaticCall,
+    "return" => AssemblyBuiltInFunction::Return,
+    "revert" => AssemblyBuiltInFunction::Revert,
+    "selfdestruct" => AssemblyBuiltInFunction::SelfDestruct,
+    "invalid" => AssemblyBuiltInFunction::Invalid,
+    "log0" => AssemblyBuiltInFunction::Log0,
+    "log1" => AssemblyBuiltInFunction::Log1,
+    "log2" => AssemblyBuiltInFunction::Log2,
+    "log3" => AssemblyBuiltInFunction::Log3,
+    "log4" => AssemblyBuiltInFunction::Log4,
+    "chainid" => AssemblyBuiltInFunction::ChainId,
+    "basefee" => AssemblyBuiltInFunction::BaseFee,
+    "origin" => AssemblyBuiltInFunction::Origin,
+    "gasprice" => AssemblyBuiltInFunction::GasPrice,
+    "blockhash" => AssemblyBuiltInFunction::BlockHash,
+    "coinbase" => AssemblyBuiltInFunction::CoinBase,
+    "timestamp" => AssemblyBuiltInFunction::Timestamp,
+    "number" => AssemblyBuiltInFunction::Number,
+    "difficulty" => AssemblyBuiltInFunction::Difficulty,
+    "gaslimit" => AssemblyBuiltInFunction::GasLimit,
+};
+
+pub fn parse_builtin_keyword(keyword: &str) -> Option<&AssemblyBuiltInFunction> {
+    BUILTIN_ASSEMBLY_FUNCTIONS.get(keyword)
+}
+
+impl AssemblyBuiltInFunction {
+    pub(crate) fn get_prototype_info(self) -> &'static AssemblyBuiltinPrototype {
+        let index = self as usize;
+        &ASSEMBLY_BUILTIN[index]
+    }
+}
+
+// Assembly built-in functions.
+// Descriptions copied and slightly modified from: https://docs.soliditylang.org/en/v0.8.12/yul.html
+static ASSEMBLY_BUILTIN: Lazy<[AssemblyBuiltinPrototype; 76]> = Lazy::new(|| {
+    [
+        AssemblyBuiltinPrototype{
```

Comment:
So the `AssemblyBuiltInFunction` is an index into this array. If there is anything misplaced, then we will get strange results. I would suggest adding a field to `AssemblyBuiltinPrototype` of type `AssemblyBuiltInFunction`. That way you can do an `assert_eq!()` or `debug_assert_eq!()` to make sure there nothing off in the array. It does cost a tiny amount of memory of course.
What do you think?

Reply:
That is a good idea. Another way of solving this would be creating a direct mapping between the keyword and the built-in prototype. This way, I can carry only static references to such item throughout the code. That will save me memory and I can eliminate the enum. What do you think of this approach?

Reply:
This is a nice solution. You may still want to have the extra field, so we can do codegen dependant on the builtin type.

Reply:
Only if needed of course.

Reply:
Either way, we need check to verify that. So I'll follow your suggestion.